### PR TITLE
Tx status query not waiting for refunds

### PR DESCRIPTION
**Pull request by anthony-near**
Fix for [issue 6834](https://github.com/near/nearcore/issues/6834). 

1. Get `receipt` from `self.store` using the `receipt_id`
2. Determine whether `receipt` is a refund by calling `receipt.predecessor_id.is_system()` See [docs](https://nomicon.io/DataStructures/Account?highlight=system#system-account) for why this works. 
3. Only include the `receipt` in the `results` if it is not a refund receipt
4. add test in https://github.com/near/nearcore/tree/master/chain/jsonrpc/jsonrpc-tests

### CODE REVIEW

**Review by mina86 — APPROVED** [review]

**Comment by matklad** [comment]
Can we write a test here?

**Comment by anthony-near** [comment]
> Can we write a test here?

Where would be a good place to put the test? looking in `chain/chain/src/tests` I don't see a clear place for such a test. 

And for the test itself, I was going to create 3 receipts, one of them being gas_refund and assert len == 2. Any more conditions to cover?

**Comment by matklad** [comment]
>Where would be a good place to put the test? 

That is a good question. I personally still didn't form a good mental map for where the tests should go for anything outside of contract runtime. 

I *think* what we care here is not so much this internal method of `Chain`, but rather the `tx_status` RPC, the ideal test I'd think would check the RPC layer. 

I think we have some RPC tests here: 

https://github.com/near/nearcore/tree/master/chain/jsonrpc/jsonrpc-tests

But I am not sure how alive that infra is. 

I think I've also seen something similar in our standard test-cases (an issue with some pointers https://github.com/near/nearcore/issues/6724), but overall, I don't know. 

@mm-near @mina86 @firatNEAR any ideas on where's the place to add new tests for these kinds of things?

**Comment by anthony-near** [comment]
Talked with Firat and he said I can add a new test at the rpc endpoint level that differentiates between refund & non-refund receipts. 

> > Where would be a good place to put the test?
> 
> That is a good question. I personally still didn't form a good mental map for where the tests should go for anything outside of contract runtime.
> 
> I _think_ what we care here is not so much this internal method of `Chain`, but rather the `tx_status` RPC, the ideal test I'd think would check the RPC layer.
> 
> I think we have some RPC tests here:
> 
> https://github.com/near/nearcore/tree/master/chain/jsonrpc/jsonrpc-tests
> 
> But I am not sure how alive that infra is.
> 
> I think I've also seen something similar in our standard test-cases (an issue with some pointers #6724), but overall, I don't know.
> 
> @mm-near @mina86 @firatNEAR any ideas on where's the place to add new tests for these kinds of things?

**Comment by anthony-near** [comment]
@matklad tried adding a test in https://github.com/near/nearcore/tree/master/chain/jsonrpc/jsonrpc-tests, but no receipts have been returned when calling `client.EXPERIMENTAL_tx_status` (even when waiting 5 seconds after broadcasting the tx). See most recent commit. Is the `test_with_client!` macro I'm using go for multiple blocks or is it only 1 block? I couldn't tell from looking at the fn definition.

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs [hunk 1]
```diff
@@ -97,6 +98,34 @@ fn test_send_tx_commit() {
...
+        let bytes = tx.try_to_vec().unwrap();
+        client.broadcast_tx_commit(to_base64(&bytes)).await.unwrap();
+        thread::sleep(time::Duration::from_secs(5));
+        let tx_status = client.EXPERIMENTAL_tx_status(to_base64(&bytes)).await.unwrap();
+        for receipt in tx_status.get("receipts") {
+            if !receipt.as_array().unwrap().is_empty() {
+                let is_not_refund = !receipt.get("predecessor_id").is_none();
+                assert!(is_not_refund);
+            }
+        }
```
Comment (matklad) [inline chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs:125]: Hm, does this actually test the desired behavior? I seem to always get an empty array of recipts, both with and without the change?

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs [hunk 2]
```diff
@@ -97,6 +98,34 @@ fn test_send_tx_commit() {
...
+            1,
+            "test1".parse().unwrap(),
+            "test2".parse().unwrap(),
+            &signer,
+            100,
+            block_hash,
+        );
+        let bytes = tx.try_to_vec().unwrap();
+        client.broadcast_tx_commit(to_base64(&bytes)).await.unwrap();
+        thread::sleep(time::Duration::from_secs(5));
```
Comment (matklad) [inline chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs:118]: Sleeping for five seconds is pretty bad, as (multiplied by the number of tests), it makes the whole test suite very slow. 

Sadly, I think some polling is required architecturally at the moment, and the best way to do that is to write a loop with a short poll interval.

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs (hunk 1 above)
Comment (anthony-near) [reply chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs]: Forgot the `is_system`. Added in updated commit.

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs (hunk 2 above)
Comment (anthony-near) [reply chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs]: polling every 10 ms for 10 times (100ms total). Happy to change those values

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs [hunk 3]
```diff
@@ -97,6 +99,43 @@ fn test_send_tx_commit() {
...
+            1,
+            "test1".parse().unwrap(),
+            "test2".parse().unwrap(),
+            &signer,
+            100,
+            block_hash,
+        );
+        let bytes = tx.try_to_vec().unwrap();
+        client.broadcast_tx_commit(to_base64(&bytes)).await.unwrap();
+        for _ in 1..10 {
```
Comment (matklad) [inline chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs:119]: If I read this correctly, this loops always runs for the ten iterations, we never break from it, right? So this seems to be equivalent to just waiting for 100 millis.

I think what we want here is to loop until we do get a response, and then break. 

And we should assert that we actually did get a non-empty response in the end.

Comment (anthony-near) [reply chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs:120]: updated to break when we get a response and asserting we get a non-empty response at the end

**Review by jakmeier — APPROVED** [review]
looks good to me, I left an optional comment how you could maybe make the test a bit nicer but I'm not sure if we have access to the relevant type there, so feel free to ignore if it's not trivial to change

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -3297,6 +3297,12 @@ impl Chain {
         let receipt_ids = outcome.outcome.receipt_ids.clone();
         let mut results = vec![outcome];
         for receipt_id in &receipt_ids {
+            if let Some(receipt) = self.store.get_receipt(&receipt_id)? {
+                let is_refund = receipt.predecessor_id.is_system();
```
Comment (mm-near) [inline chain/chain/src/chain.rs:3301]: please add a comment here that we don't include the refund receipts on purpose.

(and please also add the comment to the function itself)

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs [hunk 5]
```diff
@@ -97,6 +99,50 @@ fn test_send_tx_commit() {
...
+                let receipt_predecessor_id = receipt.get("predecessor_id");
+                if receipt_predecessor_id.is_some() {
+                    let is_refund = receipt["predecessor_id"].get("is_system").unwrap().as_bool();
+                    if is_refund.is_some() {
+                        assert!(!is_refund.unwrap());
+                    }
+                }
+            }
+        }
+        assert!(tx_status.get("receipts").unwrap().as_array().is_some());
```
Comment (mm-near) [inline chain/jsonrpc/jsonrpc-tests/tests/rpc_transactions.rs:142]: could we check that number  of receipts is matching what we expect?

**Review by mm-near — APPROVED** [review]

**Pull request merged**